src/gausskernel/storage/access/k2/catalog/sql_catalog_manager.h

```diff
@@ -0,0 +1,183 @@
+    The TableInfoHandler is used to acccess the TableInfo (including IndexInfo) objects on SKV
+
+All tables in a database are mapped to a SKV collection by database id, not database name to make the
+rename database working. Similary, a table's SKV schema name is identified by the table id, not the table name to
+avoid moving data around after renaming a table.
+
+To store a TableInfo with possible multiple IndexInfo on SKV (an IndexInfo represents a secondary index for a table),
+we have the following three catalog tables with fixed schema defined in table_info_handler.h for each database
+1) table meta: store table and index informaton. One entry for a table or an index
+2) table column meta: store column information for a table. One entry for a table column
+3) index column meta: store index column for an index. One entry for an index column
+
+The TableInfoHandler is used to persist or read a TableInfo/IndexInfo object on SKV. When a TableInfo is stored to the
+above catalog tables, the actual SKV schema is obtained dynamically and is created by TableInfoHandler as well by
+table id so that DMLs could insert, update, select, and delete from the SKV table.
+
+Please be aware that blocking APIs are used for the catalog manager for now. Will refactor them to include task submission
+and task state checking APIs later by using thread pools.
+
+*/
+#pragma once
+
+#include "cluster_info_handler.h"
+#include "database_info_handler.h"
+
+namespace k2pg {
+namespace sql {
+namespace catalog {
+struct CreateDatabaseRequest {
+    std::string databaseName;
+    std::string databaseId;
+    uint32_t databaseOid;
+    std::string sourceDatabaseId;
+    std::string creatorRoleName;
+    // next oid to assign. Ignored when sourceDatabaseId is given and the nextPgOid from source database will be used
+    std::optional<uint32_t> nextPgOid;
+};
+
+class SqlCatalogManager  {
+public:
+    SqlCatalogManager();
+    ~SqlCatalogManager();
+
+    sh::Status Start();
+
+    void Shutdown();
+
+    // use synchronous APIs for the first attempt here
+    // TODO: change them to asynchronous with status check later by introducing threadpool task processing
+    sh::Status InitPrimaryCluster();
+
+    sh::Status FinishInitDB();
+
+    sh::Response<bool> IsInitDbDone();
+
+    uint64_t  GetCatalogVersion();
+
+    sh::Response<uint64_t> IncrementCatalogVersion();
+
+    sh::Response<std::shared_ptr<DatabaseInfo>> CreateDatabase(const CreateDatabaseRequest& request);
+
+    sh::Response<std::vector<DatabaseInfo>> ListDatabases();
+
+    sh::Response<std::shared_ptr<DatabaseInfo>> GetDatabase(const std::string& databaseName, const std::string& databaseId);
+
+    sh::Status DeleteDatabase(const std::string& databaseName, const std::string& databaseId);
+
+    sh::Status UseDatabase(const std::string& databaseName);
+
+protected:
+    std::atomic<bool> initted_{false};
+
+    mutable std::mutex lock_;
```

Comment:
Not sure about this lock and the atomic variables. Is the intent that there will be only one instance of this class for the whole process? It might make more sense to have one instance per session/thread and then you don't need any of this synchronization overhead. Maybe you can discuss with Ivan and Jian next week.

Reply:
Let us discuss this later.